### PR TITLE
Load apt.dat from custom scenery

### LINKED ISSUE #254: Load airports from custom scenery
- Stop reading the big apt.dat from Resources/default scenery/default apt dat
- Read apt.dat only according to scenery_packs.ini (which includes the big apt.dat in "Custom Scenery/Global Scenery/" folder)
- Skip duplicate airports - first wins according to the order in scenery_packs.ini 

Note: we will have 170 airports less - those which were only found in the apt.dat from Resources/default scenery/default apt dat
On the other hand, loading both big apt.dat files would take 1 minute.
So currently there doesn't seem to be an alternative.

### PR DESCRIPTION
**Pull request by crankywright**
Fixes #254 

What I do in this PR:
- Stop reading the big apt.dat from Resources/default scenery/default apt dat
- Read apt.dat only according to scenery_packs.ini (which includes the big apt.dat in "Custom Scenery/Global Scenery/" folder)
- Skip duplicate airports - first wins according to the order in scenery_packs.ini 

Now we have 170 airports less - those which were only found in the apt.dat from Resources/default scenery/default apt dat
On the other hand, loading both big apt.dat files would take 1 minute.
So currently there doesn't seem to be an alternative.

### CODE REVIEW

**Comment by crankywright** [comment]
@Random-Ranger 
Mentioning you in case you want to peer-review this pull request.

**Review by Random-Ranger — APPROVED** [review]
Everything looks good to me

**Review by Random-Ranger — COMMENTED** [review]
@felix-b 
I made some tests to see if we could load the global "apt.dat" after the files referenced from scenery_pack.ini. I thought it should be quite fast because we have las than 200 airports to build from this file but found out that it took another 30 seconds. Quite disappointing.

I made the test on Windows and on linux in a WSL Ubuntu Vm, that's what made me find the problem with LastNonSpaceIndex (I was reading the Windows file from linux and the \r at the end of the line in the scenery_pack.ini was a problem).

I have no good solution for the first problem but I suspect the file IOs to be the limiting factor here.
In fact, I made a quick and dirty hack, replacing the ifstream with a stringstream which loads all the file in memory before starting to parse it and the loading of the nearly 30000 airports went from 30s to 15s. This is probably not the right solution because that means loading 350MB of data in memory but there is some room for improvement here.

File: src/libloaders/tokenValueFileReaderBase.cpp [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        catch(const exception &e)
+        {
+            break;
+        }
+
+        const char *delimiterChars = ",: ";
+        size_t delimitierIndex = text.find_first_of(delimiterChars);
+        int lastNonSpaceIndex = delimitierIndex == string::npos ? -1 : (int)text.length() - 1;
+
+        while (lastNonSpaceIndex >= 0 && strchr(delimiterChars, text.at(lastNonSpaceIndex)))
```
[the remark this answers is not in the document]
Comment (crankywright) [reply src/libloaders/tokenValueFileReaderBase.cpp:32]: @Random-Ranger 
Good catch, thanks - will fix

**Comment by crankywright** [comment]
@Random-Ranger 

> I made some tests to see if we could load the global "apt.dat" after the files referenced from scenery_pack.ini. I thought it should be quite fast because we have las than 200 airports to build from this file but found out that it took another 30 seconds. Quite disappointing.

There's no way to read only specific airports from apt.dat without passing through the entire file line by line. This is one of the reasons for using a binary cache. Or we can create a binary index that maps airport ICAOs to file paths and offsets + lengths in the files - although if we go for the cache, why bother with apt.dat indexing.

> I have no good solution for the first problem but I suspect the file IOs to be the limiting factor here.
In fact, I made a quick and dirty hack, replacing the ifstream with a stringstream which loads all the file in memory before starting to parse it and the loading of the nearly 30000 airports went from 30s to 15s. This is probably not the right solution because that means loading 350MB of data in memory but there is some room for improvement here.

I think you got it. We need a large buffer.
I'll try increasing the ifstream buffer: https://en.cppreference.com/w/cpp/io/basic_streambuf/pubsetbuf
I'll try 16MB or 32MB and see what I get

**Comment by crankywright** [comment]
@Random-Ranger 
No breakthrough so far, as `pubsetbuf()`  which I tried to use is implementation-specific and in my case (MinGW) it had no effect.
I can fiddle with providing a custom `streambuf`, or alternatively, custom buffering logic.
But I'd rather do that as part of a dedicated optimization effort - which is going to happen at some point, for sure

**Comment by Random-Ranger** [comment]
@crankywright 

I know we have to read the whole file but I was thinking, naively, that skipping nearly all the airports in the file should be way faster than reading and building all the airports.

I agree that it is not the point of this issue to optimize the loading.

**Pull request merged**